src/libertem/analysis/clust.py

```diff
@@ -24,47 +25,76 @@ class ClusterTemplate(GeneratorHelper):
-        controller = controller.split("\n\n")
-        temp_controller = []
-        controller[1] = controller[1].replace("self", "cluster_analysis")
-        controller[4] = controller[4].replace("self.", "")
-        controller[7] = controller[7].replace("self.dataset", "ds")
-        controller[2] = "sd_udf_results = ctx.run_udf(dataset=ds,\
-                         udf=stddev_udf, roi=roi, progress=True)"
-        temp_controller.extend(controller[1:9])
-        result = ["udf_results = ctx.run_udf(dataset=ds, udf=udf, progress=True)",
-                  "cluster_result = cluster_analysis.get_udf_results\
-                  (udf_results=udf_results, roi=roi)"]
-        temp_controller.append("\n".join(result))
-        return '\n\n'.join(temp_controller)
+        temp_controller = [
+                        "dataset = ds",
+                        "roi, sd_udf_results = get_sd_results()",
+                        "cluster_udf = get_cluster_udf(sd_udf_results)",
+                        "udf_results = ctx.run_udf(dataset=ds, udf=cluster_udf, progress=True)",
+                        "cluster_result = get_udf_results(udf_results, roi)",
+        ]
+        return '\n'.join(temp_controller)
 
     def get_docs(self):
         docs = ["# Cluster Analysis"]
         return '\n'.join(docs)
 
+    def get_run_sd_udf(self):
+        indent = " "*4
+        temp_run_sd = [
+            "def run_sd_udf(roi, stddev_udf):",
+            f"{indent}sd_udf_results = ctx.run_udf(dataset=ds, udf=stddev_udf, roi=roi,"
+            " progress=True)",
+            f"{indent}return roi, consolidate_result(sd_udf_results)"
+        ]
+        return "\n".join(temp_run_sd)
+
     def get_analysis(self):
-        temp_analysis = [f"parameters={self.params}",
-                         "cluster_analysis = ClusterAnalysis(dataset=ds, parameters=parameters)"]
-        temp_analysis.append(self.temp_cluster_controller())
-        return '\n'.join(temp_analysis)
+        udf_results = self.format_code(inspect.getsource(ClusterAnalysis.get_udf_results))
+        sd_roi = self.format_code(inspect.getsource(ClusterAnalysis.get_sd_roi))
+        cluster_udf = self.format_code(inspect.getsource(ClusterAnalysis.get_cluster_udf))
+        sd_results = self.format_code(inspect.getsource(ClusterAnalysis.get_sd_results))
+        run_sd_udf = self.get_run_sd_udf()
+        cluster_controller = self.temp_cluster_controller()
+
+        temp_analysis = [
+                    f"parameters={self.params}\ndataset=ds",
+                    f"{sd_roi}",
+                    f"{run_sd_udf}",
+                    f"{sd_results}",
+                    f"{cluster_udf}",
+                    f"{udf_results}",
+                    f"{cluster_controller}"
+                    ]
+        return '\n\n'.join(temp_analysis)
 
     def get_plot(self):
         plot = ["plt.figure()",
                 "plt.imshow(cluster_result['intensity'].visualized)"]
```

Comment:
I think here it would be better to `imshow` the raw results, such that the user can inspect the underlying values when plotting interactively.

Reply:
Yes in principle. However, the analysis currently has an inconsistency where the raw data is not available. Maybe we can make that a FIXME for now?

Reply:
> the analysis currently has an inconsistency where the raw data is not available.

I didn't get it, what is the inconsistency here?

Reply:
> Yes in principle. However, the analysis currently has an inconsistency where the raw data is not available. Maybe we can make that a FIXME for now?

Hm, as far as I can see, the `AnalysisResultSet` returned from the clustering analysis was changed here to return the labels as `raw_data` - wouldn't that work for visualization in the notebook?

Reply:
Ah I see, I didn't notice that it was changed already. :-) Yes, that will work!